packages/tldraw/src/lib/ui/components/Toolbar/OverflowingToolbar.tsx

```diff
@@ -0,0 +1,199 @@
+import { preventDefault, useEditor, useEvent } from '@tldraw/editor'
+import classNames from 'classnames'
+import hotkeys from 'hotkeys-js'
+import { createContext, useEffect, useId, useLayoutEffect, useMemo, useRef, useState } from 'react'
+import { PORTRAIT_BREAKPOINT } from '../../constants'
+import { useBreakpoint } from '../../context/breakpoints'
+import { areShortcutsDisabled } from '../../hooks/useKeyboardShortcuts'
+import { TLUiToolItem } from '../../hooks/useTools'
+import { useTranslation } from '../../hooks/useTranslation/useTranslation'
+import { TldrawUiButton } from '../primitives/Button/TldrawUiButton'
+import { TldrawUiButtonIcon } from '../primitives/Button/TldrawUiButtonIcon'
+import {
+	TldrawUiDropdownMenuContent,
+	TldrawUiDropdownMenuRoot,
+	TldrawUiDropdownMenuTrigger,
+} from '../primitives/TldrawUiDropdownMenu'
+import { TldrawUiMenuContextProvider } from '../primitives/menus/TldrawUiMenuContext'
+
+export const IsInOverflowContext = createContext(false)
+
+export function OverflowingToolbar({ children }: { children: React.ReactNode }) {
+	const editor = useEditor()
+	const id = useId().replace(/:/g, '_')
+	const breakpoint = useBreakpoint()
+	const msg = useTranslation()
+
+	const overflowIndex = Math.min(8, 5 + breakpoint)
+
+	const [totalItems, setTotalItems] = useState(0)
+	const mainToolsRef = useRef<HTMLDivElement>(null)
+	const [lastActiveOverflowItem, setLastActiveOverflowItem] = useState<string | null>(null)
+
+	const css = useMemo(() => {
+		const showInMainSelectors = []
+		const hideFromOverflowSelectors = []
+
+		if (lastActiveOverflowItem) {
+			showInMainSelectors.push(`[data-value="${lastActiveOverflowItem}"]`)
+		} else {
+			showInMainSelectors.push(`:nth-child(${overflowIndex + 1})`)
```

Comment:
i'm curious: is there a way to do this without the dynamic css?
like, can there be a demarcated element with a className that looks at a boundary or some other kind grouping?

Reply:
i don't _think_ so? steve's branch was trying that - we can sort of make it work for truncating the list, but not really for the active item which is pretty dynamic. open to suggestions tho!